Fix #195: Acceptance tests based on Cucumber

hferentschik opened the pull request:
To execute the Cucumber tests:

    $ bundle install
    $ bundle exec rake get_adb
    $ bundle exec rake get_cdk
    $ bundle exec rake features 

Reports in _build/features_report.html_.

Alternatively one can manually copy the box files to the expected directory (see README)

I switched from vagrant-spec to Cucumber, since the former seems to be lest maintained and the syntax and reporting of Cucumber/Gherkin is much better.

Thoughts?

brgnepal commented on README.md:
```diff
@@ -98,23 +100,60 @@ Exit Code Number   | Meaning
+    $ bundle exec rake -T
+
+<a name="acceptance-tests"></a>
+## Acceptance tests
```
I think `Acceptance Tests` would be more appropriate here.

hferentschik replied:
Why? You mean a capital 'T'? I don't think so?

brgnepal commented on README.md:
```diff
@@ -98,23 +100,60 @@ Exit Code Number   | Meaning
+
+The tests assume that the ADB and CDK box files are available under
+_build/boxes/adb-\<provider\>.box_ resp _build/boxes/cdk-\<provider\>.box_. You can
+either copy the box files manually or use the _get_adk_ resp. _get_cdk_ Rake tasks.
```
Typo error here `_get_adk_`

hferentschik replied:
thanks

brgnepal commented on README.md:
```diff
@@ -98,23 +100,60 @@ Exit Code Number   | Meaning
+_build/boxes/adb-\<provider\>.box_ resp _build/boxes/cdk-\<provider\>.box_. You can
+either copy the box files manually or use the _get_adk_ resp. _get_cdk_ Rake tasks.
+
+After test execution the Cucumber test reports can be found under _build/features_report.html_
```
It is possible add a task like `bundle exec rake view_reports` which will spawn browser opening file `build/features_report.html`. It will greatly enhance user experience. 
Right now, it looks like I need to manually open this html file.

hferentschik replied:
Not quite sure how this enhances the user experience, but I'll add one. Note, there is more typing involved opening the browser via Rake instead of directly. 

In case you want to display everything in the console you can also execute `bundle exec rake features:pretty`

brgnepal commented on Rakefile:
```diff
@@ -1 +1,95 @@
+require 'fileutils'
+require 'yaml'
+
+CDK_DOWNLOAD_URL='https://access.redhat.com/downloads/content/293/ver=2/rhel---7/2.0.0/x86_64/product-software'
```
Probably better to extract these links in separate file says `constants` in order to reduce `Rakefile` size.

hferentschik replied:
-1 I don't think the Rakefile size is an issue for now. The more files the more files the harder it becomes to piece everything together. 

A valid argument might be to extract this into some sort of properties file to be able to change these settings w/o affecting the _Rakefile_, but I don't think it is worth the effort here. I would suggest we wait and see how this shapes up.

budhrg commented on Rakefile:
```diff
@@ -1 +1,95 @@
+namespace :features do
+  cucumber_config = YAML.load_file('./.config/cucumber.yml')
+  cucumber_config.each_key do |key|
+    if key == 'default'
```
One liner is preferred here as `next if key == 'default'`

brgnepal replied:
One liner is preferred here as `next if key == 'default'`

hferentschik replied:
sure why not

brgnepal commented on Rakefile:
```diff
@@ -1 +1,95 @@
+
+desc 'Download CDK Vagrant box using the specified provider (default \'virtualbox\')'
+task :get_cdk, [:provider] do |t, args|
+  if args[:provider].nil?
```
Probably ternary operator is better here.
`provider = args[:provider].nil? ?  'virtualbox' : args[:provider]`

hferentschik replied:
matter of taste, but why not

brgnepal commented on Rakefile:
```diff
@@ -1 +1,95 @@
+  else
+    provider=args[:provider]
+  end
+  agent = Mechanize.new
```
I would prefer to take out these lower level operations in separate file and then call it here.

Adding these code here actually make the `Rakefile` bulky and would be difficult to read it as a whole when it scales up.

hferentschik replied:
So where do you see it scaling to? I don't consider 92 lines of _Rakefile_ as bulky and I am not su sure whether it will grow so much.

brgnepal commented on Rakefile:
```diff
@@ -1 +1,95 @@
+  else
+    provider=args[:provider]
+  end
+  agent = Mechanize.new
```
Same as this comment https://github.com/projectatomic/vagrant-service-manager/pull/205/files#r62502746

brgnepal commented on Rakefile:
```diff
@@ -1 +1,95 @@
+
+desc 'Download ADB Vagrant box using the specified provider (default \'virtualbox\')'
+task :get_adb, [:provider] do |t, args|
+  if args[:provider].nil?
```
Same as this comment https://github.com/projectatomic/vagrant-service-manager/pull/205/files#r62501910

brgnepal commented:
@hferentschik Implementation looks great :+1: . Need to test it locally.

brgnepal commented on README.md:
```diff
@@ -1,34 +1,36 @@
-* [Builds](#builds)
+<!-- MarkdownTOC -->
 
+- [Objective](#objective)
```
@bexelbie @Preeticp Please have a look at this new TOC with nested feature.

brgnepal commented on Gemfile.lock:
```diff
@@ -0,0 +1,179 @@
+GIT
```
It is not recommended to check-in `Gemfile.lock` while building gem. 
Check here http://stackoverflow.com/a/4151540/1120530

hferentschik replied:
I am aware of this discussion and I don't necessarily agree. It will depend very much on your workflow and whether or not you regularly delete your Gemfile.lock. If not, one can detect a potential gem conflict only on the first checkout and 'bundle install'. 

Personally, I'd rather ensure that we have a consistent dev environment and do dedicated bundle upgrades/checks as prep work for a release. 

Alternatively, one sets up a dedicated CI jobs which only job it is to test each commit against a fresh 'bundle install' (aka deleting the lock file). If it fails, someone can have a look and fix the issue. On the plus side, each developer checking out the repository and trying to build it will have a consistent experience. This might give you the best of both worlds.

hferentschik replied:
Also, we are talking about "locking" the dev dependencies here. There are really no runtime deps.

brgnepal replied:
Updated my comment here https://github.com/projectatomic/vagrant-service-manager/issues/203#issuecomment-218162976.

hferentschik commented:
@budhrg thanks for the review.

I fixed the typo and made some cosmetic changes to the _Rakefile_ and added a _features:open_report_ open report task.

Regarding the _Gemfile.lock_, I think I still prefer to lock the version for development and mitigate gem file incompatibilities via a CI job and dedicated version checks prior to releases. If we agree on the other changes, we can drop the last commit for now and apply the test changes first in order to move on.

hferentschik commented:
> Implementation looks great :+1: . 

thanks

> Need to test it locally.

Let me know how it goes.

hferentschik commented:
@dharmit so what you I need to do now to get access to a CentOS CI server?

hferentschik commented:
Ok, I removed the checkin of the _Gemfile.lock_, we can take this discussion later. I also added support for testing against _libvirt_ when available.

Maybe we can get  this merged now. After all this is "just" a test framework. We can improve/adjust over time. I just don't won't to have to rebase forever.

brgnepal commented on README.md:
```diff
@@ -98,23 +100,68 @@ Exit Code Number   | Meaning
+_build/boxes/adb-\<provider\>.box_ resp _build/boxes/cdk-\<provider\>.box_. You can
+either copy the box files manually or use the _get_adb_ resp. _get_cdk_ Rake tasks.
+
+Per default the VirtualBox provider is used. However, you can also run against Libvirt
```
Can we rewrite it as Virtualbox provider is used as default provider.

hferentschik replied:
Why, the default spelling is VirtualBox as far as I know.

brgnepal replied:
I mean to say Let's rewrite this whole sentence as:

```
VirtualBox is used as a default provider. However, you can also run against Libvirt
```

As starting sentence using `Per default` doesn't sound good to me. However, I would like @bexelbie  and @Preeticp to review this README since they are mostly handling docs for ADB.

brgnepal commented on README.md:
```diff
@@ -98,23 +100,68 @@ Exit Code Number   | Meaning
+After test execution the Cucumber test reports can be found under _build/features_report.html_.
+They can also be opened via
+
+    $ bundle exec rake features:open_report
```
:+1:

hferentschik replied:
your wish is my command :-)

LalatenduMohanty commented on README.md:
```diff
@@ -98,23 +100,68 @@ Exit Code Number   | Meaning
-You can contact us on:
-  * IRC: #atomic and #nulecule on freenode
-  * Mailing List: container-tools@redhat.com
+After cloing the repository, install the [Bundler](http://bundler.io/) gem:
```
Spell correction s/cloing/cloning/

hferentschik replied:
thanks

LalatenduMohanty commented on Rakefile:
```diff
@@ -1 +1,91 @@
+    # Submit the login form
+    after_login = login_page.form_with(:name => 'login_form') do |f|
+     username_field = f.field_with(:id => 'username')
+     username_field.value = 'service-manager@mailinator.com'
```
Is it a real email ID?

brgnepal replied:
Its a disposable email id provided by `https://mailinator.com/`

hferentschik replied:
Right. Avoids to pass anything in via environment variables. Works out of the box.

brgnepal commented on Rakefile:
```diff
@@ -1 +1,91 @@
+  cucumber_config = YAML.load_file('./.config/cucumber.yml')
+  cucumber_config.each_key do |key|
+    next if key == 'default'
+    Cucumber::Rake::Task.new(key.to_sym, "Run Cucumber features using profile '#{key}' (see cucumber.yml)'") do |t|
```
Probably `see cucumber.yml` should be `see .config/cucumber.yml`. I had hard time to see `cucumber.yml` file then went to PR and found it is inside hidden folder `.config`.

hferentschik replied:
it's the default directory of the cucumber config.

hferentschik commented:
Rebased once more. AFAIU, you are seeing issues on Fedora? I see whether I can help. That said, it just works fine out of the box on OS X and Linux Mint. 

I'd still suggest to merge this as is to be able to move on. You guys can create new issues with some error reports on what the problem is on lovely Fedora :S

hferentschik commented:
BTW, how important would be the ability to execute the tests on Windows? Obviously one would need to change some of the expectations on Windows (eg the evaluable output), but the big rub seems to be that the Cucumber approach won't work on Windows. The reason being are several bugs in Aruba :-(

Not quite sure what to say here atm. From a syntax and functionality point of view I would favor Cucumber, but if we want to have a reasonable shot of getting this to work on Windows we would be back to vagrant-spec.

hferentschik commented:
For a reference, here is the bug report for running Aruba on Windows - https://github.com/cucumber/aruba/issues/387. Seems the project is still quite active and there is definitely interest to fix this. It might be ok to go with the Cucmber + Aruba approach under the assumption/hope that there will be a fix to the problem soonish.

brgnepal commented on README.md:
```diff
@@ -118,15 +126,75 @@ Exit Code Number   | Meaning
+    # Run tests against CDK using Libvirt
+    $ bundle exec rake features BOX=cdk PROVIDER=libvirt
+    
+    # Run against ADB and CDK
```
probably `Run against ADB and/or CDK`. As one can give one or both names.

hferentschik replied:
Sure, but the example is for the 'AND' case

hferentschik replied:
Made a small change

brgnepal commented on README.md:
```diff
@@ -118,15 +126,75 @@ Exit Code Number   | Meaning
+
+You can also run a single feature specifying the explicit feature file to use:
+
+    $ bundle exec rake features FEATURE=features/<feature-filename>.feature
```
:+1:

hferentschik commented:
Pushed an update which allows to mix and match which features are run against which provider and box usinf env variables _FEATURE_, _BOX_ and _PROVIDER_. See README. Also made IP part of the test scenario

bexelbie commented on README.md:
```diff
@@ -118,15 +126,75 @@ Exit Code Number   | Meaning
+    $ bundle exec rake features:open_report
+
+<a name="builds"></a>
+## Builds
```
This section can be deleted, imho.

hferentschik replied:
yeah :-) I wanted to ask about this. I do it right away!

bexelbie commented:
README LGTM with note about possible duplicative section deletion.

brgnepal commented:
@hferentschik it looks like you took the oldest commit while squashing.

Date of commit is `April 25` now. 

commit 1e65f76046aa7fa66cf7dc8bc75359bb17d71821
Author: Hardy Ferentschik hardy@hibernate.org
Date:   Mon Apr 25 14:43:26 2016 +0200

brgnepal commented on README.md:
```diff
@@ -118,15 +125,69 @@ Exit Code Number   | Meaning
+    $ bundle exec rake features BOX=cdk PROVIDER=libvirt
+
+    # Run against ADB and CDK (boxes are comma seperated)
+    $ budle exec rake features BOX=cdk,adb
```
Typo error @hferentschik

hferentschik replied:
thanks

LalatenduMohanty commented on README.md:
```diff
@@ -118,15 +125,69 @@ Exit Code Number   | Meaning
+The tests assume that the ADB and CDK box files are available under
+_build/boxes/adb-\<provider\>.box_ resp _build/boxes/cdk-\<provider\>.box_. You can
+either copy the box files manually or use the _get_adb_ resp. _get_cdk_ Rake tasks.
+
```
Is there a way where I can use aleady installed Vagrant box for the tests? For example I should be able able to use `config.vm.box = "projectatomic/adb"`

LalatenduMohanty replied:
This will save some internet bandwidth for me as I dont have  to re-download box again which is already on the workstation i.e Fedora. I had downloaded the box using following commands

```
vagrant init projectatomic/adb
vagrant up
```

The Vagrant box is kept inside /home/lmohanty/.vagrant.d/boxes/projectatomic-VAGRANTSLASH-adb/2.0.0 extracted format i.e

```
[lmohanty@LalatenduM-laptop 2.0.0]$ pwd
/home/lmohanty/.vagrant.d/boxes/projectatomic-VAGRANTSLASH-adb/2.0.0

[lmohanty@LalatenduM-laptop 2.0.0]$ tree ./
./
└── libvirt
    ├── box.img
    ├── metadata.json
    └── Vagrantfile

1 directory, 3 files
```

brgnepal replied:
@LalatenduMohanty you can convert that into `.box` file using Vagrant [`package`](https://www.vagrantup.com/docs/cli/package.html) command.

It goes like this:

```
1. mkdir tmp_adb
2. vagrant init projectatomic/adb
3. vagrant up
4. vagrant package --vagrantfile Vagrantfile --output ~/Downloads/adb-virtualbox.box
```

My first attempt was through those local boxes but it seems that those are corrupt([fpaste here](http://paste.fedoraproject.org/365264/14629788/)).

We can create an issue to figure that out.

LalatenduMohanty replied:
But why we cant just reuse the installed box? I do not think repackaging the Vagrant box with `vagrant package --vagrantfile Vagrantfile --output ~/Downloads/adb-virtualbox.box` is a good idea.

brgnepal replied:
That's why I added at the bottom that we can figure it out later all possible options :smile: .

LalatenduMohanty replied:
Looks like the actually adds the box, un-compresses  the box file , hence the issue

brgnepal commented:
LGTM :+1: 

```
➜  vagrant-service-manager git:(hardy_issue_195) ✗ bundle exec rake features BOX=adb,cdk
/home/budhram/.rvm/rubies/ruby-2.2.1/bin/ruby -S bundle exec cucumber 
Using the default and html profiles...
...................--------.--------.........................................-------------------.-------------------.......................----------.----------...........---------...............------.------
18 scenarios (9 skipped, 9 passed)
208 steps (95 skipped, 113 passed)
13m28.174s
```

LalatenduMohanty commented:
The CDK tests on libvirt worked fine for me . Trying out it on OSX

```
[lmohanty@LalatenduM-laptop vagrant-service-manager]$ bundle exec rake features BOX=cdk PROVIDER=libvirt
Your Gemfile lists the gem vagrant-service-manager (>= 0) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.

/usr/bin/ruby -S bundle exec cucumber 
Using the default and html profiles...
.-----------------.........---------.---------------------------------------....................--------------------.---------------------...........-----------.---------...........-------------.......-------

18 scenarios (13 skipped, 5 passed)
208 steps (146 skipped, 62 passed)
4m27.067s
```

brgnepal commented:
> The CDK tests on libvirt worked fine for me

@LalatenduMohanty Could you run tests with both?

```
bundle exec rake features BOX=adb,cdk PROVIDER=libvirt
```

LalatenduMohanty commented:
> Could you run tests with both?

I don't have vagrant box for ADB on my local workstation (the way test wants it) and downloading going to take time :(

brgnepal commented:
@LalatenduMohanty ok will check with ADB - libvirt combination and merge this PR.

LalatenduMohanty commented:
Got a ADB 1.8.0 box and it seems to work

```
[lmohanty@LalatenduM-laptop vagrant-service-manager]$ bundle exec rake features BOX=adb PROVIDER=libvirt
Your Gemfile lists the gem vagrant-service-manager (>= 0) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.

/usr/bin/ruby -S bundle exec cucumber 
Using the default and html profiles...
---------.-----------------.........--------------------.---------------------------------------....................-----------.---------------------...........---------------------------.-------------.......

18 scenarios (14 skipped, 4 passed)
208 steps (157 skipped, 51 passed)
2m46.741s
```

hferentschik commented:
> Is there a way where I can use aleady installed Vagrant box for the tests? For example I should be able able to use config.vm.box = "projectatomic/adb"

Right now, no. It could be done, but then you need different Vagrantfile (one w/ URL and one w/o) for each box or work out. Makes the setup more complicated and you most likely need to duplicate test code. Hence I opted against it. Effectively you are getting the ADB box from Hashicorp anyways. That's where it is downloaded from, if you use the download task. 

I am also a big fan of keeping everything w/i the project, which includes tmp artifacts, tmp files and directories, etc. If I delete _pkg_ and _build_, there should be nothing left which got created by any development tasks.

> This will save some internet bandwidth for me as I dont have to re-download box again

Come on, since when do we care re-downloading something like this. I bet you you have downloaded the box more than once already. So yes, the current approach downloads and keeps the box file local to the project, but until you delete the _build_ (run clean task) directory it will be used from there.

I also think it is in general a bad idea to interact with anything in ~/.vagrant. For example, when experimenting next with moving config into the box _Vagrantfile_, I will modify the _Vagrantfile_ in ~/.vagrant. If at the same time I would run tests against the CDK using the box in ~/.vagrant I might get unexpected results. 

Also definitely a -1 for trying to re-package the box.

LalatenduMohanty commented:
> That's where it is downloaded from, if you use the download task.

It is not exactly the same. Basically if I use atlas I get un-compressed  box. But if I use cloud.c.o I get the whole box.

LalatenduMohanty commented:
`bundle install` giving error on OS X as I think because gemfile has fog-libvirt

```
$ bundle install
Your Gemfile lists the gem vagrant-service-manager (>= 0) more than once.
You should probably keep only one of them.
While it's not a problem now, it could cause errors if you change the version of just one of them later.
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 11.1.1


Your user account isn't allowed to install to the system Rubygems.
You can cancel this installation and run:

    bundle install --path vendor/bundle

to install the gems into ./vendor/bundle/, or you can enter your password
and install the bundled gems to Rubygems using sudo.

Password: 
Installing addressable 2.4.0
Using ffi 1.9.10
Using childprocess 0.5.9
Installing contracts 0.14.0
Using builder 3.2.2
Installing gherkin 3.2.0
Installing cucumber-core 1.4.0
Installing cucumber-wire 0.0.1
Using diff-lcs 1.2.5
Using multi_json 1.11.2
Installing multi_test 0.1.2
Installing cucumber 2.3.3
Installing rspec-support 3.4.1
Installing rspec-expectations 3.4.0
Installing thor 0.19.1
Installing aruba 0.14.1
Using bundler 1.8.0
Using unf_ext 0.0.7.2
Using unf 0.1.4
Using domain_name 0.5.20160309
Using erubis 2.7.0
Installing events 0.9.8
Installing excon 0.49.0
Installing formatador 0.2.5
Installing fog-core 1.39.0
Installing fog-json 1.0.2
Using mini_portile2 2.0.0
Using nokogiri 1.6.7.1
Installing fog-xml 0.1.2
Using json 1.8.3

Gem::Installer::ExtensionBuildError: ERROR: Failed to build gem native extension.

    /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby extconf.rb 
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
    --with-opt-dir
    --without-opt-dir
    --with-opt-include
    --without-opt-include=${opt-dir}/include
    --with-opt-lib
    --without-opt-lib=${opt-dir}/lib
    --with-make-prog
    --without-make-prog
    --srcdir=.
    --curdir
    --ruby=/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
    --with-libvirt-include
    --without-libvirt-include
    --with-libvirt-lib
    --without-libvirt-lib
    --with-libvirt-config
    --without-libvirt-config
    --with-pkg-config
    --without-pkg-config
extconf.rb:73:in `<main>': libvirt library not found in default locations (RuntimeError)


Gem files will remain installed in /var/folders/81/q3z841ds2yb7xblc91xww10m0000gn/T/bundler20160513-5213-1xln3oc/ruby-libvirt-0.6.0/gems/ruby-libvirt-0.6.0 for inspection.
Results logged to /var/folders/81/q3z841ds2yb7xblc91xww10m0000gn/T/bundler20160513-5213-1xln3oc/ruby-libvirt-0.6.0/gems/ruby-libvirt-0.6.0/ext/libvirt/gem_make.out
An error occurred while installing ruby-libvirt (0.6.0), and Bundler cannot continue.
Make sure that `gem install ruby-libvirt -v '0.6.0'` succeeds before bundling.
```

navidshaikh commented:
Can we open a separate ticket for issues we are finding for OS X and get this PR merged for Linux platform? This should unblock us to have a CI job first and we keep improving the tests for other platforms as well?

LalatenduMohanty commented:
@navidshaikh lets do that.

navidshaikh commented:
Well, I should have merged using command line, this github UI merge will add one more commit in history. :-(  Next time!

hferentschik commented:
LOL, thanks anyway

hferentschik commented:
This button can be so tempting ;-)
